streaming/base/format/base/reader.py

```diff
@@ -114,51 +130,78 @@ def init_local_dir(self, filenames_present: Set[str], keep_zip: bool) -> bool:
         # the shard raw as not present, in order to achieve consistency.
         if not raw_files_present:
-            is_raw_present = False
+            has_raw = False
         elif raw_files_present < len(self.file_pairs):
-            is_raw_present = False
+            has_raw = False
             self._evict_raw()
         else:
-            is_raw_present = True
+            has_raw = True
 
         # Same as the above, but for shard zip files.
         if not zip_files_present:
-            is_zip_present = False
+            has_zip = False
         elif zip_files_present < len(self.file_pairs):
-            is_zip_present = False
+            has_zip = False
             self._evict_zip()
         else:
-            is_zip_present = True
-
-        # Do we keep_zip?
-        if keep_zip:
-            # If we can keep_zip, and we do, and have either raw or zip, we must have the other one
-            # too, because they are downloaded and decompressed together.
-            if self.compression and (is_zip_present != is_raw_present):
-                if is_raw_present:
-                    is_raw_present = False
-                    self._evict_raw()
-                elif is_zip_present:
-                    is_zip_present = False
-                    self._evict_zip()
+            has_zip = True
+
+        # Enumerate cases of raw/zip presence.
+        if self.compression:
+            if safe_keep_zip:
+                if has_raw:
+                    if has_zip:
+                        # Present (normalized).
+                        pass
+                    else:
+                        # Missing: there is no natural way to arrive at this state, so drop raw.
+                        has_raw = False
+                        self._evict_raw()
+                else:
+                    if has_zip:
+                        # Present: but missing raw, so need to decompress upon use.
+                        pass
+                    else:
+                        # Missing (normalized).
+                        pass
+            else:
+                if has_raw:
+                    if has_zip:
+                        # Present: zip is unnecessary, so evict it.
+                        has_zip = False
+                        self._evict_raw()
+                    else:
+                        # Present (normalized).
+                        pass
+                else:
+                    if has_zip:
+                        # Present: but missing raw, so need to decompress and evict zip upon use.
+                        pass
+                    else:
+                        # Missing (normalized).
+                        pass
         else:
-            # If we don't keep_zip, drop any zip files.
-            if is_zip_present:
-                is_zip_present = False
-                self._evict_zip()
+            if has_zip:
+                raise ValueError('Shard is invalid: compression was not used, but has a ' +
```

Comment:
Is there a test for this exception?

Reply:
This would occur if someone manually edit the index.json file and make it corrupted.